File: backends/micronaut-java/src/main/java/dev/stackverse/backend/EventLog.java
package dev.stackverse.backend;

import org.slf4j.Logger;
import org.slf4j.spi.LoggingEventBuilder;

import java.util.Map;

final class EventLog {
    private EventLog() {
    }

    static void info(Logger logger, String event, String outcome, String message) {
        with(logger.atInfo(), event, outcome).log(message);
    }

    static void info(Logger logger, String event, String outcome, String message, Map<String, ?> fields) {
        LoggingEventBuilder builder = with(logger.atInfo(), event, outcome);
        fields.forEach(builder::addKeyValue);
        builder.log(message);
    }

    static void warn(Logger logger, String event, String outcome, String message, Map<String, ?> fields) {
        LoggingEventBuilder builder = with(logger.atWarn(), event, outcome);
        fields.forEach(builder::addKeyValue);
        builder.log(message);
    }

    private static LoggingEventBuilder with(LoggingEventBuilder builder, String event, String outcome) {
        return builder.addKeyValue("event", event).addKeyValue("outcome", outcome);
    }
}

File: backends/ruby-rails/app/services/stackverse/event_log.rb
module Stackverse
  module EventLog
    LEVELS = {
      debug: Logger::DEBUG,
      info: Logger::INFO,
      warn: Logger::WARN,
      error: Logger::ERROR,
      fatal: Logger::FATAL
    }.freeze

    module_function

    def log(level, event, outcome, message, fields = {})
      Rails.logger.add(
        LEVELS.fetch(level, Logger::INFO),
        { message: message, event: event, outcome: outcome }.merge(fields.compact),
        "stackverse.backend.ruby_rails"
      )
    end

    def info(event, outcome, message, fields = {})
      log(:info, event, outcome, message, fields)
    end

    def warn(event, outcome, message, fields = {})
      log(:warn, event, outcome, message, fields)
    end

    def error(event, outcome, message, fields = {})
      log(:error, event, outcome, message, fields)
    end
  end
end
